scrapy_poet/injection.py

```diff
@@ -0,0 +1,323 @@
+import inspect
+import logging
+from typing import Dict, Callable, Any, List, Set
+
+from twisted.internet.defer import inlineCallbacks, returnValue
+
+import andi
+from scrapy import Request, Spider
+from scrapy.crawler import Crawler
+from scrapy.http import Response
+from scrapy.settings import Settings
+from scrapy.statscollectors import StatsCollector
+from scrapy.utils.defer import maybeDeferred_coro
+from scrapy.utils.misc import load_object
+from scrapy_poet.injection_errors import (UndeclaredProvidedTypeError,
+                                          NonCallableProviderError,
+                                          InjectionError)
+from scrapy_poet.page_input_providers import PageObjectInputProvider, PROVIDERS
+from scrapy_poet.api import _CALLBACK_FOR_MARKER, DummyResponse
+from web_poet.pages import is_injectable
+
+
+logger = logging.getLogger(__name__)
+
+
+class Injector:
+    """
+    Keep all the logic required to do dependency injection in Scrapy callbacks.
+    Initializes the providers from the spider settings at initialization.
+    """
+    def __init__(self, crawler: Crawler):
+        self.crawler = crawler
+        self.spider = crawler.spider
+        self.load_providers()
+
+    def load_providers(self):
+        self.providers = [
+            cls(self.crawler)
+            for cls in load_provider_classes(self.spider.settings)
+        ]
+        check_all_providers_are_callable(self.providers)
+        # Caching whether each provider requires the scrapy response
+        self.is_provider_requiring_scrapy_response = {
+            id(provider): is_provider_requiring_scrapy_response(provider)
```

Comment:
Why is `id` call needed here?

Reply:
If the idea is to avoid making a reference, https://docs.python.org/3/library/weakref.html#weakref.WeakKeyDictionary can be better.

Reply:
Keeping references is not a problem (the provider's instances are kept internally within the injector). The main reason is avoiding expensive operations if the provider implements an overridden `__eq__` (or even if it was wrongly implemented). For this particular case, it is known that there will be a single instance of each provider and it can be identified by the id well. Why would WeakKeyDictionary be better? 

In other words, for example, I want to avoid using the expensive comparator that `attr.s` could create for a provider if somebody is using it.

Reply:
For me `id` is always suspicious because Python can reuse ids between different object, and actually does so by default in many cases. So, an object is created, you take an id from it, then the object is deallocated (e.g. nobody refers to it anymore), a completely different object is created, and it may get the same id. You then do a lookup and get a wrong value, misteriously.

This is not a theoretical issue, as it happens by default for some objects (I think tuples? not sure) - Python is free to use preallocated pools, and reuse pointers of objects which no longer have references, to avoid extra memory allocation. 

I don't think this happens here, but the code just doesn't feel "correct" - it may lead to very hard-to-debug bugs. You take id(proider), which doesn't create a new reference to provider, so nothing prevents provider to be gone, and id to be reused. If you use provider itself as a key, this can't happen. It also can't happen with weakref.

Reply:
I see your point. Sure that Python is reusing memory, at the very minimum as C does: If you call `malloc` after `free` you could get the same memory address. But - correct me if I'm wrong - this won't happen for a living object: once the object is alive, it will retain the same address. In fact, the default `__eq__` implementation for `object` just uses `id()` (again correct me if I'm wrong). 

>  You take id(proider), which doesn't create a new reference to provider, so nothing prevents provider to be gone

Well, it is impossible for the provider to be gone. Only if somebody externally modifies the internal list of providers in the `Injector`... but note that in this case, this will also fail with a `WeakKeyDictionary`: a newly introduced provider won't be present in the dictionary and the `Injector` will just fail. The current implementation doesn't allow for run-time updating of the providers by design. We could do it, but It doesn't seem required. 

So well, as conclusion, I think using `id()` is an optimization and in this case is not risky (I'm just forcing using `is` vs `==` for comparison), but it is also Ok to me to remove the `id` and just use the instance. Using `WeakKeyDictionary` doesn't provide any benefit in this case IMHO. Implemented in: https://github.com/scrapinghub/scrapy-poet/pull/41/commits/4fed87ae81f9e58e6aaa64b69bc28afce11ab352

Reply:
Yeah, that's fair that for living objects id is fine. I guess my issue is that with `id(provider)` code becomes non-local: if you use provider as a key, it guarantees that the object will be alive (as a mapping now has a reference to the object); with id it is not guaranteed, so you you have to read the whole program to know if using `id(provider)` is valid or not here. 

It is like a very light https://en.wikipedia.org/wiki/COMEFROM :) in a sense global code structure needs to be checked to know what happens locally.

Reply:
TIL COMEFROM :smile: